keps/sig-storage/2848-transfer-volumesnapshots-namespaces/README.md

```diff
@@ -0,0 +1,801 @@
+
+#### Story 1
+
+- The `prod` namespace contains a PVC that holds user information. VolumeSnapshots are taken of
+this PVC on a regular basis.
+
+- An existing VolumeSnapshot can be transferred to a test namespace. The user initiates the request
+by creating a `StorageTransferRequest` in the source namespace.
+
+- Once created, a `StorageTransferApproval` resource in the target namespace will be created.
+If matching resources are found, the VolumeSnapshot controller can init the transfer. 
+
+- When the transfer is successful, a new VolumeSnapshotContent is being created one-to-one mapping between VolumeSnapshot and VolumeSnapshotContent.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+This KEP proposes the minimal to transfer VolumeSnapshot between namespaces, while avoiding breaking the current API.
+Many caveats have been discussed, including but not limited to :- 
+
+- [StorageClass Secrets](https://kubernetes-csi.github.io/docs/secrets-and-credentials-storage-class.html).
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+A `StorageTransferRequest` named `r` and `StorageTransferApproval` `a` are matched if all of the following conditions are met:
+
+- `r.metadata.name == a.spec.requestName`
+- `r.metadata.namespace == a.spec.targetNamespace`
+- `r.spec.source.namespace == a.metadata.namespace`
+- `r.spec.source.name == a.spec.source.name`
+- `r.spec.source.kind == a.spec.source.kind`
+- `r.spec.approvalName == a.metadata.name`
+
+Once matching is found, the transfer process is initiated by the controller, and it performs the following:
+
+- Changes source VolumeSnapshot `ReadyToUse` to `false`, so it can not be used.
+- Changes VolumeSnapshotContent deletion policy to retain.
+- Creates a VolumeSnapshot in the target namespace.
+- Creates a new VolumeSnapShotContent.  
+- Binds the newly created VolumeSnapshotContent to the newly created VolumeSnapshot.
+- Changes original VolumeSnapshotContent deletion policy to delete.  
```

Comment:
> Changing the deletion policy of the source snapshot is dangerous an unnecessary.

Or, we should add a kind of field to `volumeSnapShotContent` to let the external-snapshotter know that it is transferred `volumeSnapShotContent` and the actual backend volume for it should not be deleted, regardless of the deletion policy (and modify external-snapshotter handle the field properly). Or, we might need to go back to the approach to allow `VolumeSnapshotRef` conditionally.

Reply:
> > Changing the deletion policy of the source snapshot is dangerous an unnecessary.
> 
> Or, we should add a kind of field to `volumeSnapShotContent` to let the external-snapshotter know that it is transferred `volumeSnapShotContent` and the actual backend volume for it should not be deleted, regardless of the deletion policy (and modify external-snapshotter handle the field properly). Or, we might need to go back to the approach to allow modifying `VolumeSnapshotRef` conditionally.

@mkimuram Thanks a lot for raising this here. 

So, I would try to avoid the second approach, which add a field to `volumeSnapShotContent`. Afaik, `VolumeSnapshotRef` is immutable.

Reply:
> I think that before `Deletes the source VolumeSnapshot.`, `VolumeSnapshotContent` policy need to be `retain`, to avoid the actual backend volume for the new VolumeSnapShotContent to be deleted.
> 
> So, should we remove `Changes original VolumeSnapshotContent deletion policy to delete.` here and move '`Changes VolumeSnapshotContent deletion policy to retain.` above, with adding "original", here?

I think after the new `volumeSnapshot` and `volumeSnapshotContent` are being created in the target namespace, we no longer need the original ones. 

I change the policy of the original one before the transfer to avoid deleting `original volumeSnapshotContent`. After the transfer is done, I change the policy back to `retain` and then delete the original `volumeSnapshot` , which consequently should deletes the original `volumeSnapshotContent` 

Please correct me if i am wrong

Reply:
According to this [document](https://kubernetes.io/docs/concepts/storage/volume-snapshots/#delete),
> Delete
> 
> Deletion is triggered by deleting the VolumeSnapshot object, and the DeletionPolicy will be followed. If the DeletionPolicy is Delete, then the underlying storage snapshot will be deleted along with the
> VolumeSnapshotContent object. If the DeletionPolicy is Retain, then both the underlying snapshot and VolumeSnapshotContent remain.

And the [code](https://github.com/kubernetes-csi/external-snapshotter/blob/c4a2ac7753c7245454cabddf935ac5c9020d8161/pkg/sidecar-controller/snapshot_controller.go#L60) actually calls the CSI `DeleteSnapshot`, which deletes the the __underlying snapshot__, if the `DeletionPolicy == crdv1.VolumeSnapshotContentDelete`.

On the other hand, current KEP describes the flow as follows:

> - Creates a new VolumeSnapShotContent.
> - Binds the newly created VolumeSnapshotContent to the newly created VolumeSnapshot.
> - Changes original VolumeSnapshotContent deletion policy to delete.
> - Deletes the source VolumeSnapshot.

So, it seems that it "Deletes the source VolumeSnapshot" with its "VolumeSnapshotContent deletion policy" set to be "delete". Doesn't it make the __underlying snapshot__ to be deleted, along with the original VolumeSnapshotContent and VolumeSnapshot?

(And the __underlying snapshot__ should be the actual backend volume that the "new VolumeSnapShotContent" is referencing to, therefore above behavior won't be what we want).